pilz_control/test/unittest_cartesian_speed_monitor.cpp

```diff
@@ -0,0 +1,255 @@
+
+  robot_state::RobotStatePtr state2(new robot_state::RobotState(model_));
+  state2->setToDefaultValues();
+  double angular_displacement{ 0.1 };
+  state2->setVariablePosition(joint_names_[0], angular_displacement);
+  state2->updateLinkTransforms();
+
+  moveit::core::LinkModel* link = model_->getLinkModel("link2");
+  double time_delta = 0.4;
+  double expected_velocity = angular_displacement / time_delta;
+
+  double speed = CartesianSpeedMonitor::linkSpeed(state, state2, link, time_delta);
+  EXPECT_NEAR(speed, expected_velocity, SPEED_COMPARISON_TOLERANCE);
+
+  link = model_->getLinkModel("link3");
+  speed = CartesianSpeedMonitor::linkSpeed(state, state2, link, time_delta);
+  EXPECT_NEAR(speed, expected_velocity, SPEED_COMPARISON_TOLERANCE);
+}
+
+/**
+ * @tests{Monitor_Speed_of_all_links_until_TCP,
+ * Tests speed monitoring of all moveable links.
+ * }
+ */
+TEST_F(CartesianSpeedMonitorTest, testLinkSpeedMoveJoint2)
+{
+  robot_state::RobotStatePtr state(new robot_state::RobotState(model_));
+  state->setToDefaultValues();
+  state->updateLinkTransforms();
+
+  robot_state::RobotStatePtr state2(new robot_state::RobotState(model_));
+  state2->setToDefaultValues();
+  double angular_displacement{ 0.1 };
+  state2->setVariablePosition(joint_names_[1], angular_displacement);
+  state2->updateLinkTransforms();
+
+  moveit::core::LinkModel* link = model_->getLinkModel("link2");
+  double time_delta = 0.3;
+
+  double speed = CartesianSpeedMonitor::linkSpeed(state, state2, link, time_delta);
+  EXPECT_NEAR(speed, 0.0, SPEED_COMPARISON_TOLERANCE);
+
+  link = model_->getLinkModel("link3");
+  double expected_velocity = angular_displacement / time_delta;
+  speed = CartesianSpeedMonitor::linkSpeed(state, state2, link, time_delta);
+  EXPECT_NEAR(speed, expected_velocity, SPEED_COMPARISON_TOLERANCE);
+}
+
+/**
+ * @tests{Monitor_Speed_of_all_links_until_TCP,
+ * Tests speed monitoring of all moveable links.
+ * }
+ */
+TEST_F(CartesianSpeedMonitorTest, testLinkSpeedLinearInTime)
+{
+  robot_state::RobotStatePtr state(new robot_state::RobotState(model_));
+  state->setToDefaultValues();
+  state->updateLinkTransforms();
+
+  robot_state::RobotStatePtr state2(new robot_state::RobotState(model_));
+  state2->setToDefaultValues();
+  std::vector<double> angular_displacements{ 0.1, 0.1 };
+  state2->setVariablePositions(angular_displacements);
+  state2->updateLinkTransforms();
+
+  moveit::core::LinkModel* link = model_->getLinkModel("link3");
+  double time_delta = 0.1;
+
+  double speed = CartesianSpeedMonitor::linkSpeed(state, state2, link, time_delta);
+
+  double factor = 2.0;
+  time_delta *= factor;
+  double speed2 = CartesianSpeedMonitor::linkSpeed(state, state2, link, time_delta);
+
+  EXPECT_NEAR(speed2 * factor, speed, SPEED_COMPARISON_TOLERANCE);
+}
+
+/**
+ * @tests{Monitor_Speed_of_all_links_until_TCP,
+ * Tests speed monitoring of all moveable links.
+ * }
+ */
+TEST_F(CartesianSpeedMonitorTest, testBelowLimitMoveJoint2)
```

Comment:
Refactor test name and test docu, it is not very meaningful because right now I would ask: Why are there two tests (`testBelowLimitMoveJoint1`and `testBelowLimitMoveJoint2`), basically doing the same thing? => Explain in more detail your test intention.

Reply:
I refactored the tests.